### DISCUSSION TITLE
Key creation with ED25519PH - missing OID

### DISCUSSION THREAD

**Discussion by commel** [openssl/openssl#23441] [Q&A] [answered]
I would like to create a key with ED25519 using ED25519PH. I've found that OpenSSL 3.3.x is missing the [OID-Definition](http://www.oid-info.com/get/1.3.101.114), but is there a way to circumvense this?

Also should this be implemented in upstream? I would be happy to provide a Pull Request if needed.

**Comment by levitte** [comment]
You can either use `OBJ_create()` to create it programmatically, or use the OpenSSL config file, like this:

https://www.openssl.org/docs/man3.2/man5/config.html#ASN.1-Object-Identifier-Configuration

What method you choose depends on how you create that key; program or openssl command line...

**Comment by commel** [comment]
This seems to add the information about the OID to the openssl context. But does this switch the OID of the generated ed25519 key to the OID of the ed25519ph?

**Reply by levitte** [reply]
Er...  no?  Purely looking at the key type, it's Ed25519 regardless of the signature scheme you use it with (Ed25519, Ed25519-ctx, or Ed25519-ph).

So I wouldn't expect to see Ed25519-ph in a AlgorithmIdentifier of a SubjectPublicKeyInfo (see [RFC8410 - Subject Public Key Fields](https://datatracker.ietf.org/doc/html/rfc8410#section-4) or PrivateKeyInfo (see [RFC8410 - Private Key Format](https://datatracker.ietf.org/doc/html/rfc8410#section-7)) structure.
I *could*, however, see such an OID in a Certificate's signatureAlgorithm (see [RFC 8410 - EdDSA Signagures](https://datatracker.ietf.org/doc/html/rfc8410#section-6))

**Reply by levitte** [reply, marked as answer]
Something to be noted is that id-EdDSA25519-ph and id-EdDSA448-ph never made it to RFC 8410, or even to the draft (the link to which you find in the [OID definition](http://www.oid-info.com/get/1.3.101.114) you pointed at) that lead to that RFC.  This pretty much correlates with the strong admonition in [RFC 8032 - Choice of Signature Primitive](https://datatracker.ietf.org/doc/html/rfc8032#section-8.5) (*emphasis mine*):

>    The Ed25519ph and Ed448ph variants are prehashed.  This is mainly
>    useful for interoperation with legacy APIs, since in most of the
>    cases, either the amount of data signed is not large or the protocol
>    is in the position to do digesting in ways better than just
>    prehashing (e.g., tree hashing or splitting the data).  The
>    prehashing also makes the functions greatly more vulnerable to
>    weaknesses in hash functions used.  *These variants SHOULD NOT be
>    used*.

... so I would be cautious about using them at all, unless you really have to.

**Reply by commel** [reply]
Thank you so much for these insights! I was not totally sure about the intended use but this clarifies this! Can you hint me in the direction on where to implement the ASN1 addition, if necessary and wanted I would like to try to implement this as Pull Request.

**Reply by stokito** [reply]
A good article on the topic [EdDSA, Ed25519, Ed25519-IETF, Ed25519ph, Ed25519ctx, HashEdDSA, PureEdDSA, WTF?](https://cryptologie.net/article/497/eddsa-ed25519-ed25519-ietf-ed25519ph-ed25519ctx-hasheddsa-pureeddsa-wtf/)
